Enforce purity on missing assigns clause

feliperodri opened the pull request:
CBMC currently doesn't enforce assigns clauses if there isn't one.
However, whenever enforcing a function contract, the absence of an
assigns clause should be interpreted as an empty write set
(i.e., no inputs will be modified by this function).
CBMC now properly checks for empty assigns clause behavior.

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

SaswatPadhi commented on regression/contracts/assigns_enforce_05/test.desc:
```diff
@@ -1,9 +1,10 @@
+^EXIT=0$
 ^SIGNAL=0$
-^VERIFICATION FAILED$
+^VERIFICATION SUCCESSFUL$
```
Is there any way we could add some additional regexes to make sure that contracts are processed? The enforce flag should be generating some additional assertions, so can we match some regex on them?

The idea here is to fail this test (instead of vacuously passing) if somehow `--enforce-all-contracts` flag is a no-op.

feliperodri replied:
Implemented as suggested.

SaswatPadhi commented on regression/contracts/assigns_enforce_15/test.desc:
```diff
@@ -0,0 +1,10 @@
+--enforce-all-contracts
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
```
Can we add some regexes to match on the assertions that failed?

feliperodri replied:
Implemented as suggested.

SaswatPadhi commented on regression/contracts/assigns_enforce_15/main.c:
```diff
@@ -0,0 +1,20 @@
+  return *a;
+}
+
+int main()
```
Can we add another function that doesn't have assigns clause, but modifies a global variable?
We can check for the appropriate assertion in the `test.desc` file instead of creating a separate regression test just for this case.

feliperodri replied:
Implemented as suggested.

SaswatPadhi commented:
I just realized that the PR doesn't actually add support for `__CPROVER_assigns()`. It only checks for purity of functions when assigns clause is missing. So may be we can rename the PR (and update the commit message) to: "Enforce purity on missing assigns clause"?

kroening commented:
> I just realized that the PR doesn't actually add support for `__CPROVER_assigns()`. It only checks for purity of functions when assigns clause is missing. So may be we can rename the PR (and update the commit message) to: "Enforce purity on missing assigns clause"?

Yes, renaming the PR would help!

TGWDB commented on regression/contracts/assigns_enforce_15/test.desc:
```diff
@@ -0,0 +1,12 @@
+^EXIT=10$
+^SIGNAL=0$
+^\[bar.1\] line 13 .*: FAILURE$
+^\[baz.1\] line 18 .*: FAILURE$
```
This looks like it will fail, isn't it line 19? I think other tests use regex like:
```
^\[main.enum-range-check.6\] line \d+ enum range check in ev3: SUCCESS$
^\[main.enum-range-check.7\] line \d+ enum range check in ev2: FAILURE$
```
taken from `regression/cbmc/enum_is_in_range/enum_test5.desc`

feliperodri replied:
Thanks, @TGWDB! Implemented as suggested.

TGWDB commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -513,20 +513,8 @@ void code_contractst::instrument_call_statement(
-  }
-    else // Called function has assigns clause
-    {
+  if(!called_assigns.is_nil()) // Called function has assigns clause
+  {
```
The alignment of the old code now inside this conditional is off by two spaces. I don't know if this is something clang-format dislikes, but it would be nice to fix.

feliperodri replied:
Implemented as suggested.

TGWDB commented:
I think this looks ok. There are some small tidying bits I noted, and I think it will fail regression tests (but this should just be fixing the test regexs). The only bit that is unclear to me is whether removing the old code that used to enforce failure is safe and won't break expected functionality. (Why was it written if this is fairly clearly failing when there is no assignment if this should instead be pure and not evaluated as a failure?)

SaswatPadhi reviewed (approved):
Approving since I don't have any additional comments, but let's fix all of the comments from @TGWDB.

feliperodri committed:
Support for empty assigns clause

CBMC currently doesn't enforce assigns clauses if there isn't one.
However, whenever enforcing a function contract, the absence of an
assigns clause should be interpreted as an empty write set
(i.e., no inputs will be modified by this function).
CBMC now properly checks for empty assigns clause behavior.

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

feliperodri commented:
> The only bit that is unclear to me is whether removing the old code that used to enforce failure is safe and won't break expected functionality. (Why was it written if this is fairly clearly failing when there is no assignment if this should instead be pure and not evaluated as a failure?)

@TGWDB Before, when enforcing a contract for function foo, CBMC would fail if any of its inner functions don't have an assigns clause, even if the inner function preserve all inputs. This behavior is too restrict. If a function doesn't change any of the inputs, its assigns clause should be empty. On this PR, we do one step further and instead of adding a false assertion for functions without assigns clause, we check for purity of the inner functions (i.e., they don't change any of the inputs).

chrisr-diffblue commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -634,9 +622,6 @@ bool code_contractst::add_pointer_checks(const std::string &function_name)
   exprt assigns_expr = type.assigns();
-  // Return if there are no reference checks to perform.
-  if(assigns_expr.is_nil())
-    return false;
```
Just a check - is there a test case that will fail if this code ever gets inserted again? :-) It superficially looks like an "obvious" optimisation, so I can imaging someone in the future re-adding it with good intent.

feliperodri replied:
If somebody add this back, `assigns_enforce_15` will fail. For this case, we have assignments, but because there is no assigns clause, we wouldn't even complain about them. we should report an error in this case, because we want to enforce contracts across all functions using `--enforce-all-contracts` flag.

chrisr-diffblue reviewed (approved):
Looks good to me, thanks - with just one small sanity check on my part :-)
